File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-kt/src/main/kotlin/io/hotmoka/network/thin/client/models/signatures/MethodSignatureModel.kt
package io.hotmoka.network.thin.client.models.signatures

/**
 * The model of the signature of a method of a class.
 */
class MethodSignatureModel(
        /**
         * The name of the method.
         */
        val methodName: String,
        /**
         * The return type of the method, null if it's void.
         */
        val returnType: String?,
        formals: List<String>,
        definingClass: String
): CodeSignatureModel(formals, definingClass)

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-js/src/models/signatures/MethodSignatureModel.ts
import {CodeSignatureModel} from "./CodeSignatureModel";
import {MarshallingContext} from "../../internal/marshalling/MarshallingContext";
import {HotmokaException} from "../../internal/exception/HotmokaException";

/**
 * The model of the signature of a method of a class.
 */
export abstract class MethodSignatureModel extends CodeSignatureModel {
    /**
     * The name of the method.
     */
    methodName: string


    constructor(methodName: string,
                definingClass: string,
                formals: Array<string>) {
        super(definingClass, formals)

        if (!methodName) {
            throw new HotmokaException("Invalid methodName " + methodName)
        }

        this.methodName = methodName
    }

    public equals(other: any): boolean {
        return other instanceof MethodSignatureModel &&
            (other as MethodSignatureModel).methodName === this.methodName &&
            super.equals(other)
    }

    public into(context: MarshallingContext): void {
        super.into(context)
        context.writeString(this.methodName)
    }
}